src/coreclr/gc/gc.cpp

```diff
@@ -23562,7 +24282,21 @@ void gc_heap::equalize_promoted_bytes()
         // check we didn't screw up the data structures
         for (int i = 0; i < n_heaps; i++)
         {
-            g_heaps[i]->verify_regions (gen_idx, false);
+            gc_heap* hp = g_heaps[i];
+
+            generation* gen = hp->generation_of (gen_idx);
+
+            // if the number of heaps changes, some heaps may end up with
+            // no regions at all - repair this here
+            if (heap_segment_rw (generation_start_segment (gen)) == nullptr)
+            {
+                size_t size = gen_idx > max_generation ? global_region_allocator.get_large_region_alignment() : 0;
+                heap_segment* start_region = hp->get_free_region (gen_idx, size);
```

Comment:
this can fail - at the minimum we should add a TODO here.

Reply:
I have added a comment.

Reply:
I rearranged the code so we check before whether we have or can get enough regions for each generation **before** we commit to increasing the heap count.